### DISCUSSION TITLE
Flexible OSAL configuration

### DISCUSSION THREAD

**Discussion by TheLancashireman** [hathach/tinyusb#1268] [Ideas]
The OSAL selection could benefit from using a construct like:

`#ifdef TUSB_OSAL_HEADER
#include TUSB_OSAL_HEADER
#else
... existing selection
#endif`

(except on separate lines, which github doesn't seem to understand)

Also: I've written an OSAL  for OSEK that should work with any OSEK-compatible OS (including AUTOSAR). I'd like to submit it for inclusion.

**Comment by hathach** [comment]
you can use the existing `OPT_OS_CUSTOM`

```C
#elif CFG_TUSB_OS == OPT_OS_CUSTOM
  #include "tusb_os_custom.h" // implemented by application
```

your suggestion is good, though, some of the code does check `CFG_TUSB_OS == OPT_OS_NONE` to see if RTOS is used to have appropriate mutex therefore the CFG_TUSB_OS should also be set to not none as well.

**Reply by TheLancashireman** [reply]
Yes, I'm using OPT_OS_CUSTOM in my own family of operating systems. But then I have to provide a tusb_os_custom.h file that uses another level of differentiation to select the right support. It would be nice to get rid of the extra level.
See https://github.com/TheLancashireman/davros/blob/master/davroska/h/tusb_os_custom.h  (location might change in the future)

Any thoughts about the OSEK OSAL?

**Reply by hathach** [reply]
> Any thoughts about the OSEK OSAL?

Yeah, please make an PR to add `OPT_OS_OSEK` and `osal_osek.h` similar to FREERTOS. All PRs are welcome. We could have external header included as you suggested though it is still not clear to me why you prefer to have the header in external location ? I am not familiar with osek, could you give a bit of explanation.
